### PR TITLE
Research/nonblocking socket

### PR DESCRIPTION
**Pull request by JuanFranDevo**
### Added
 * Sender: NonBlocking Socket support added
 * `responses` dependency open from ==0.22.0 to >=0.22.0
### Fixed
 * Sender: Closing handshake protocol improved to avoid losing of events. Upstream channel is closed and then client waits for the closing of downstream channel 
 * Sender: Monitoring of the EOF signal in downstream channel to detect whether server endpoint have closed the session
 * Fix message file management in CLI

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -47,7 +52,7 @@ in `devo/sender/data.py`
 ### Fixed
  * Ingestion endpoint has an inactivity timeout that when reached closes the connection. `devo-sdk` is aware of such a timeout and restart connection before is reached. New parameter `inactivity_timeout` in class `Sender` to set up it. Its default value is 30 seconds.
  * Syntax error when calling `Path.is_file()`
- * Documentation related to parameter `key` removal at `devo.sender.lookup.Lookup.send_data_line` in version 5.0.0
+ * Documentation related to parameter `key` removal at `devo.sender.lookup.Lookup.send_data_line` in version 5.0.0
```
Comment (oscarsierradevo) [inline CHANGELOG.md:55]: trailing space at EOL

**Review by oscarsierradevo — DISMISSED** [review]

File: tests/sender/send_data.py [hunk 2]
```diff
@@ -75,6 +76,19 @@ def test_compose_mem(self):
                          b'<14>1991-02-20 12:00:00 %s test.tag: '
                          % self.localhost.encode("utf-8"))
 
+    @staticmethod
+    def read(con, length: int):
+        if not select.select([con.socket], [], [], con.socket_timeout)[0]:
+            raise Exception("Timeout reached during read operation")
```
Comment (angel-devo) [inline tests/sender/send_data.py:82]: A generic exception here? Is there any other more concrete type to apply?

File: tests/sender/send_lookup.py [hunk 3]
```diff
@@ -70,17 +77,19 @@ def test_ssl_lookup_new_line(self):
         lookup = Lookup(name=self.lookup_name, historic_tag=None, con=con)
         p_headers = Lookup.list_to_headers(["KEY", "HEX", "COLOR"], "KEY")
         lookup.send_control("START", p_headers, "INC")
-        if len(con.socket.recv(1000)) == 0:
+        if len(TestLookup.read(con, 1000)) == 0:
             raise Exception("Not msg sent!")
```
Comment (angel-devo) [inline tests/sender/send_lookup.py:81]: Same here, I don't know if a more concrete exception applies along these lines.

File: devo/sender/data.py [hunk 4]
```diff
@@ -566,8 +579,9 @@ def close(self):
         """
         if self.socket is not None:
             try:
-                self.socket.shutdown(SHUT_RDWR)
-            except Exception:  # Try else continue
+                self.socket.shutdown(SHUT_WR)
+                self.__wait_for_EOF()
+            except Exception as exc:  # Try else continue
                 logging.exception(ERROR_MSGS.CLOSING_ERROR)
             self.socket.close()
```
Comment (angel-devo) [inline devo/sender/data.py:586]: Maybe this should be inside a `finally` clause?

```python
    def close(self):
        """
        Forces socket closure
        """
        if self.socket is not None:
            try:
                self.socket.shutdown(SHUT_WR)
                self.__wait_for_EOF()
            except Exception as exc:  # Try else continue
                logging.exception(ERROR_MSGS.CLOSING_ERROR)
            finally:
                self.socket.close()
                self.socket = None
```

File: tests/sender/send_data.py (hunk 2 above)
Comment (JuanFranDevo) [reply tests/sender/send_data.py:82]: Yes, it is just a test case

File: tests/sender/send_lookup.py (hunk 3 above)
Comment (JuanFranDevo) [reply tests/sender/send_lookup.py:81]: Yes, it is just a test case

**Review by angel-devo — APPROVED** [review]

**Pull request merged**